modules/tinymce/src/themes/silver/main/ts/ContextToolbar.ts

```diff
@@ -174,41 +176,48 @@ const register = (editor: Editor, registryContextToolbars, sink, extras) => {
 
   const getScopes: () => ScopedToolbars = Thunk.cached(() => {
     return ToolbarScopes.categorise(registryContextToolbars, (toolbarApi) => {
-      const alloySpec = buildToolbar(toolbarApi);
+      // TODO should this get multiple ctx too?
+      const alloySpec = buildToolbar([toolbarApi]);
       AlloyTriggers.emitWith(contextbar, forwardSlideEvent, {
         forwardContents: wrapInPopDialog(alloySpec)
       });
     });
   });
 
-  const buildToolbar = (ctx): AlloySpec => {
+  const buildContextToolbarGroups = (allButtons, ctx) => {
+    return identifyButtons(editor, { buttons: allButtons, toolbar: ctx.items, allowToolbarGroups: false }, extras, Option.some([ 'form:' ]));
+  };
+
+  const buildContextMenuGroups = (ctx, backstage) => {
+    return ContextForm.buildInitGroups(ctx, backstage);
+  };
+
+  const buildToolbar = (toolbars: Array<ContextTypes>): AlloySpec => {
     const { buttons } = editor.ui.registry.getAll();
+    const scopes = getScopes();
+    const allButtons = { ...buttons, ...scopes.formNavigators };
 
     // For context toolbars we don't want to use floating or sliding, so just restrict this
     // to scrolling or wrapping (default)
     const toolbarType = getToolbarMode(editor) === ToolbarMode.scrolling ? ToolbarMode.scrolling : ToolbarMode.default;
 
-    const scopes = getScopes();
-    return ctx.type === 'contexttoolbar' ? (() => {
-      const allButtons = { ...buttons, ...scopes.formNavigators };
-      const initGroups = identifyButtons(editor, { buttons: allButtons, toolbar: ctx.items, allowToolbarGroups: false }, extras, Option.some([ 'form:' ]));
-      return renderToolbar({
-        type: toolbarType,
-        uid: Id.generate('context-toolbar'),
-        initGroups,
-        onEscape: Option.none,
-        cyclicKeying: true
-      });
-    })() : (() => {
-      return ContextForm.renderContextForm(toolbarType, ctx, extras.backstage);
-    })();
+    const initGroups = Arr.flatten(Arr.map(toolbars, (ctx) => ctx.type === 'contexttoolbar' ? buildContextToolbarGroups(allButtons, ctx) : buildContextMenuGroups(ctx, extras.backstage)));
+
+    return renderToolbar({
+      type: toolbarType,
+      uid: Id.generate('context-toolbar'),
+      initGroups,
+      onEscape: Option.none,
+      cyclicKeying: true
+    });
   };
 
   editor.on(showContextToolbarEvent, (e) => {
     const scopes = getScopes();
     // TODO: Have this stored in a better structure
     Obj.get(scopes.lookupTable, e.toolbarKey).each((ctx) => {
-      launchContext(ctx, e.target === editor ? Option.none() : Option.some(e as DomElement));
+      // TODO should this do multiple ctx too?
```

Comment:
Probably not, as this is meant to open a specific toolbar IIUC? As an example we have the following in the link plugin to trigger the link context form to show:
```
editor.fire('contexttoolbar-show', {
  toolbarKey: 'quicklink'
});
```

Reply:
Yeah, was tracking that through yesterday and thought the same. Annoyingly because it's an event it could be triggered with anything? I doubt we've documented the event though, so users shouldn't trigger it. Might just leave a comment explaining this...